deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,556 @@
+    outlier_db_fn = f"{tmp_dir}/OutlierDB.pickle"
+    outlier_files = list(map(lambda x: f"{tmp_dir}/{x}.pdb", top[2]))
+    rmsds = top[3]
+    db = OutlierDB(tmp_dir, list(zip(rmsds, outlier_files)))
+    with open(outlier_db_fn, "wb") as f:
+        pickle.dump(db, f)
+    return db
+
+
+def publish(tmp_dir, published_dir):
+    """Publish outliers and the corresponding database for simulations to pick up"""
+    dbfn = f"{published_dir}/OutlierDB.pickle"
+    subprocess.getstatusoutput(f"touch {dbfn}")
+
+    mylock = LockFile(dbfn)
+
+    mylock.acquire()
+    print(
+        subprocess.getstatusoutput(
+            f"rm -rf {published_dir}/*.pdb {published_dir}/*.npy"
+        )
+    )
+    print(subprocess.getstatusoutput(f"mv {tmp_dir}/* {published_dir}/"))
+    mylock.release()
+
+    return
+
+
+def top_outliers(
+    cfg: OutlierDetectionConfig,
+    cvae_input: Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray],
+    outlier_list: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Find top num_sim outliers sorted by rmsd.
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    cvae_input : Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+            steps, positions, velocities, md5sums, rmsds
+    outlier_list : np.ndarray
+            indices corresponding to outliers
+
+    Returns
+    -------
+    Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+         Positions, velocities, md5sums, rmsds, outlier
+         indices of outliers, sorted in ascending order by rmsd
+
+    """
+    N = cfg.num_sim
+    outlier_list = list(outlier_list[0])
+    positions = cvae_input[1][outlier_list]
+    velocities = cvae_input[3][outlier_list]
+    md5s = cvae_input[2][outlier_list]
+    rmsds = cvae_input[4][outlier_list]
+
+    z = list(zip(positions, velocities, md5s, rmsds, outlier_list))
+    z.sort(key=lambda x: x[3])
+    z = z[:N]
+    z = list(zip(*z))
+
+    return z
+
+
+def main(cfg: OutlierDetectionConfig):
+    print(subprocess.getstatusoutput("hostname")[1])
+    sys.stdout.flush()
+
+    print(cfg)
+    with Timer("wait_for_input"):
+        adios_files_list = wait_for_input(cfg)
+    with Timer("wait_for_model"):
+        model_path = str(wait_for_model(cfg))
+
+    mystreams = STREAMS(
+        adios_files_list,
+        lastN=cfg.lastN,
+        config=cfg.adios_xml_agg,
+        stream_name="AggregatorOutput",
+        batch=cfg.batch,
+    )
+
+    top_dir, tmp_dir, published_dir = dirs(cfg)
+    eps = cfg.init_eps
+    min_samples = cfg.init_min_samples
+
+    # Infinite loop of outlier search iterations
+    for j in itertools.count(0):
+        print(f"outlier iteration {j}")
+
+        timer("outlier_search_iteration", 1)
+
+        with Timer("outlier_read"):
+            cvae_input = mystreams.next()
+
+        with Timer("outlier_predict"):
+            cm_predict = predict(cfg, model_path, cvae_input)
```

Comment:
Here the type of cvae_input is Tuple[List[ndarray], List[ndarray], List[ndarray], List[ndarray], List[ndarray]]

but predict definition says it takes a np.ndarray. I realize the code runs since inside the predict function it indexes into the 0th element of this tuple. But either the type should be fixed, or cvae_input[0] should be passed to the function.

Reply:
Done